File: Annotation/app/src/main/java/com/example/annotation/MainActivity.java
package com.example.annotation;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;


public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        test1();
        Class c1 = null;
        Class c2 = null;
        try {
            c1 = Class.forName("com.example.annotation.User");
            c2 = Class.forName("com.example.annotation.User");
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        User user = new User();
        System.out.println("------------------");
        System.out.println(c1.hashCode());
        System.out.println(c2.hashCode());
        System.out.println(User.class.hashCode());
        System.out.println(user.getClass().hashCode());
    }

    @MyAnnotation
    public void test() {
        System.out.print("test");
    }

    @MyAnnotation1
    public void test1() {

    }

    @MyAnnotation2("默认")
    public void test2() {

    }
}

@Target({ElementType.TYPE, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@interface MyAnnotation2 {
    String value();
}

File: RecyclerView/app/src/main/java/com/example/recyclerview/MainActivity.kt
package com.example.recyclerview

import android.content.Context
import android.content.Intent
import android.database.DatabaseUtils
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.databinding.DataBindingUtil
import androidx.databinding.ObservableField
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.recyclerview.databinding.ActivityMainBinding
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query

class MainActivity : AppCompatActivity() {
//    lateinit var mBinding: ActivityMainBinding
//    lateinit var mViewModel: ViewModel

    lateinit var mBinding: ActivityMainBinding
    lateinit var mViewModel: PaoViewModel

    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val remote = Retrofit.Builder()
            .baseUrl("http://api.jcodecraeer.com/")
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .build().create(PaoService::class.java)
        mViewModel = PaoViewModel(remote)
        mBinding.vm = mViewModel

//        setContentView(R.layout.activity_main)

//        var animal = Animal("dog", 0)
//        textView.text = "${animal.name} 叫了 ${animal.shoutCount} 声..."
//        button.setOnClickListener {
//            animal.shoutCount ++
//            textView.text  = "${animal.name} 叫了 ${animal.shoutCount} 声..."
//        }

//        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
//        val animal = Animal("dog", 0)
//        mViewModel = ViewModel(animal, this)
//        mBinding.vm = mViewModel
    }
}

data class Animal(val name: String, var shoutCount: Int)

class ViewModel(val animal: Animal, context: MainActivity) {
    val info = ObservableField<String>("${animal.name} 叫了 ${animal.shoutCount} 声...")

    fun shout(view: View) {
        animal.shoutCount ++
        info.set("${animal.name} 叫了 ${animal.shoutCount} 声...")
    }

    fun load(view: View) {

    }
}


// 服务
interface PaoService {
    @GET("article_detail.php")
    fun getArticleDetail(@Query("id") id: Int): Single<Article>
}

// 模型
class Article {

}

// vm
class PaoViewModel(val remote: PaoService) {
    val articleDetail = ObservableField<String>()

    fun loadArticle() {
        remote.getArticleDetail(8773)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ t: Article? ->
                articleDetail.set(t?.toString())
            }, { t: Throwable? ->
                articleDetail.set(t?.message ?: "error")
            })
    }
}
